### ISSUE TITLE
Why is the mode argument needed for AES deinit?

### ISSUE THREAD

**Issue by stevew817** [eembc/securemark-tls#27]
https://github.com/eembc/securemark-tls/blob/83700423e4692cf798e60820845dae200268ca99/profile/th_api/th_aes.c#L35
As the title indicates - is the mode argument really needed to deinit a structure?

**Comment by petertorelli** [comment]
It is there because Arm mbedtls has different free functions based on the type:

```C
void
th_aes_deinit(void *p_context, ee_aes_mode_t mode)
{
    if (mode == EE_AES_CCM)
    {
        mbedtls_ccm_free(&((th_mbedtls_aes_context_t *)p_context)->ctx.ccm_ctx);
    }
    else if (mode == EE_AES_GCM)
    {
        mbedtls_gcm_free(&((th_mbedtls_aes_context_t *)p_context)->ctx.gcm_ctx);
    }
}
```

wolfSSL does not require the type, however.

**Comment by stevew817** [comment]
If it is mbedTLS specific, why not store the context type in the context (`th_mbedtls_aes_context_t`) itself? As far as I'm concerned, this is polluting the generic test harness API.

**Comment by petertorelli** [comment]
That's a good point. We already do it with SHA and ECC functions, why should AES be different. Will fix!

> As far as I'm concerned, this is polluting the generic test harness API.

Exactly the kind of second-eyes input I'm looking for.

**Comment by petertorelli** [comment]
Fixed in [14da82b](https://github.com/eembc/securemark-tls/commit/14da82ba06b895f725a18547f0685f2b52375d4a). I noticed you had already added `aes_mode` to the context structure in mbedTLS so I used that.
